### ISSUE TITLE
Parallelise CTDComm to exploit multiple-GPUs and multiple-nodes

### ISSUE THREAD

**Issue by Edward-RSE** [Edward-RSE/CTDComm#5]
Following from Issue #1, we should make sure that we can use multiple GPUs to speed things up more.

https://pytorch.org/docs/stable/distributed.html
https://pytorch-geometric.readthedocs.io/en/latest/tutorial/distributed.html

**Comment by Edward-RSE** [comment]
This is actually a very complicated problem to go to a distributed training scheme. At the moment, we can take advantage of all the GPUs on a node by having parallel environments and using shared memory for the policy net (although this can be quite slow on GPUs due to synchronisation overheads!). For distributed training across nodes, we will need to use `torch.nn.parallel.DistributedDataParallel` to distribute the model across nodes and then aggregate the results into a single model.

**Comment by Edward-RSE** [comment]
OK, so the general idea of distributed training is demonstrated in this repository here: https://github.com/Edward-RSE/torch-multip/. The implementation of using torch.distributed instead of torch.multiprocessing should be straight forwards, to some degree. However, we need to be careful when thinking about how to do a distributed learning environment. I wonder if we can have one learning env per node, for example, and use all the available CPU cores and push actions to one GPU?

**Comment by Edward-RSE** [comment]
OK, so to make things easier I am going to start by assuming distributed tasks will be launched using `torchrun`. This saves me the trouble using the approach which spawns processes, because it's not exactly clean/simple to have the same `DistributedTrainer` spawns `DistributedWorkers`, where `DistributedTrainer` sends messages to tell what the workers to do. That's because torch.distributed.send/send_object_list aren't as flexible as the communication pipes in multiprocessing.

I think that the easier thing to do is to use `torchrun` to spawn the processes, then we only need the `DistributedTrainer` and not the workers it spawns. I think.

Maybe another option would be to create one `DistributedTrainer` per node, have that spawn workers and then use local communicator groups.... but that is more complicated.

**Comment by Edward-RSE** [comment]
Ah, the last point probably won't work because the ranks and spawned proc ids/ranks will be all offset to one another, so `torch.distributed.init_process_group` will never start up.
